ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +187,81 @@ public void shutdown (boolean async)
-     * @return the previous mode before attempting the suspend
+     * <p>
+     *     Make all scanning operations suspend.
+     * </p>
+     * <p>
+     *     This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
+     *     threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
+     *     progress when this method is called it will complete its scan without suspending.
+     * </p>
+     * <p>
+     *     Note that this method is also influenced by
+     *     <code>RecoveryEnvironmentBean.setWaitForFinalRecovery(boolean)</code>.
+     *     Please, refer to the its javadoc for further details.
+     *     In case <code>isWaitForFinalRecovery() == true</code>, it is <b>very important</b> that, before invoking
+     *     <code>suspendScan</code>:
+     *     * The Transaction System (i.e. TxControl) gets disabled (i.e. no new transactions will be created)
+     *     * The Transaction Reaper finishes monitoring all in-flight transactions
+     *     * All in-flight transactions without a timeout are beyond the prepare phase of the 2PC protocol;
+     *     to check when this condition is verified, com.arjuna.ats.arjuna.coordinator.ActionManager can be employed
+     * </p>
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     *              In case <code>isWaitForFinalRecovery() == true</code>, this parameter is override.
+     * @return the previous mode before attempting the suspension
      */
 
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForFinalRecovery checks whether suspension should delay until RecoveryModules
+            * implementing SuspendBlockingRecoveryModule recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
+
+               /*
+                * Now, it is finally possible to start checking if there are transactions that
+                * are still in need of recovery (or completion, in case of heuristics).
+                * The Action Manager gets checked as in-flight transactions with timeout zero
+                * can still be around (as the Transaction Reaper doesn't handle those txns)
+                */
+               while (_blockSuspension) {
```

Comment:
I'm trying to understand how _blockSuspension gets set to false: is the PeriodicRecovery thread still running at this point and if it is it will clear it eventually but then my follow up question would be why the previous statement, namely doWork(); on line 231, is called since the run loop already calls doWork()?

Reply:
Only instances of SuspendBlockingRecoveryModule can cause _blockSuspension to be set to false; if there aren't any configured recovery modules that implement SuspendBlockingRecoveryModule then this code will hang the server.

Reply:
I addressed the bug but I will keep this conversation open to discuss `doScanningWait(); doWork();`

Reply:
The design of this PR has been developed to be resilient even when external users send scan requests to the Recovery Manager. Thus, we shouldn't worry about interferences. Since we removed:
* The pre-conditions from the Javadoc, and
* References to the Action Manager

I think we can resolve this discussion. What Do You Think? @mmusgrov?